### ISSUE TITLE
Tokio runtimes and async

### ISSUE THREAD

**Issue by bendk** [mozilla/uniffi-rs#1726]
I've been thinking about the current [async_runtime](https://github.com/mozilla/uniffi-rs/blob/main/fixtures/futures/src/lib.rs#L193-L198) support and wondering how useful it is:

- If all you want to do is use tokio types and not manage any threads, it seems better to use versions of those types that are executor-agnostic like `async-mutex` and `async-timer` 
- If you want tokio to manage a threadpool, then async_compat isn't a great solution for this since it doesn't allow you to customize the tokio Runtime.

What if instead of using `async-compat`, we allowed users to specify a function that returns a type that derefs to `tokio::Runtime` and let them build their own runtime?  Something like:

```
use once_cell::sync::Lazy;
use tokio::runtime::{Builder,Runtime};

fn main_runtime() -> &'static Lazy<Runtime> {
    static RUNTIME: Lazy<Runtime> = Lazy::new(|| {
         Builder::new_multi_thread()
             .worker_threads(4)
             .thread_name("my-custom-name")
             .thread_stack_size(3 * 1024 * 1024)
             .build()
             .unwrap()
    });
    &RUNTIME
}

#[uniffi::export(tokio_runtime = main_runtime)]
pub async fn use_shared_resource(options: SharedResourceOptions) -> Result<(), AsyncError> {
    ...
}
```

Then inside the scaffolding function, we do what `async-compat` does and [enter the runtime before polling the future](https://github.com/smol-rs/async-compat/blob/master/src/lib.rs#L300-L307).

I'm not really sure how this should work.  @jplatte @Hywan are you currently using UniFFI async with tokio?  If so, how does it work?

**Comment by Hywan** [comment]
I agree that the current `async_runtime` implementation is limited but it has the merit to work.

What I don't feel clear with your proposal is: how is it supposed to work? It's focusing on defining a `Runtime`, which is a specific type of `tokio` in this case. Each async lib comes with its runtime definition. I believe the correct abstraction level is `Future`, as `async-compat` does (I'm not saying we should stick with it, but I reckon the approach is correct —though too strict/limiteed for now).

We would need a way to express we want to wrap the outgoing `Future` inside another `Future` that can be build in some way. Something like:

```rust
#[uniffi::export(async_wraps_with = future_wrapper)]
pub async fn foo(…) -> … { }

fn future_wrapper<F, T>(future: F) -> F
where F: Future<Output = T>>
{
    // … fetch `Runtime` from somewhere
    // build a new `Future` that uses `Runtime`
    // i.e. it simply is a “custom re-implementation” of `async_compat::Compat`.
}
```

Thoughts?

**Comment by jplatte** [comment]
@Hywan I don't think we should be over-abstracting things. There isn't really a use case for this outside of tokio compatibility, is there? I think there's a discussion to be had on whether tokio compat should be something UniFFI bothers with *at all*, but if it does I don't see much merit to abstracting it such that it can *theoretically* handle things other than tokio, that nobody actually needs.

**Comment by jplatte** [comment]
Well, other runtimes don't need explicit compatibility code like tokio does. The only other major runtime, `async-std`, uses a lazily-initialized global runtime so it does not need this.

**Comment by mhammond** [comment]
> My proposal allows to choose the runtime and to configure the runtime you want _per function_.

We've a vaguely-defined use-case that would like to choose the runtime dynamically at runtime (roughly, a kind of adaptor that would either want to use an async stack supplied by the foreign code or supplied by a Rust implementation, depending on the environment the code finds itself in) - @bendk is working through making that more concrete though...

**Comment by pronebird** [comment]
Instead of `#[uniffi::export(tokio_runtime = main_runtime)]` it would be great to add a new macro or as a part of scaffolding that would enable one to provide a shared tokio runtime for execution, which would automatically apply to all async functions in the module.

```rs
static SHARED_RUNTIME: LazyLock<Runtime> = ..; 

uniffi_scaffolding!(tokio_runtime = SHARED_RUNTIME);
```

**Comment by pronebird** [comment]
@bendk @mhammond  please consider making it possible to configure runtime once for the entire scaffolding
